File: core/src/main/java/com/oneandone/typedrest/ActionEndpointImpl.java
package com.oneandone.typedrest;

import java.io.*;
import java.net.*;
import org.apache.http.client.fluent.*;

/**
 * REST endpoint that represents a single RPC-like action.
 */
public class ActionEndpointImpl
        extends AbstractTriggerEndpoint
        implements ActionEndpoint {

    /**
     * Creates a new action endpoint with a relative URI.
     *
     * @param parent The parent endpoint containing this one.
     * @param relativeUri The URI of this endpoint relative to the
     * <code>parent</code>'s.
     */
    public ActionEndpointImpl(Endpoint parent, URI relativeUri) {
        super(parent, relativeUri);
    }

    /**
     * Creates a new action endpoint with a relative URI.
     *
     * @param parent The parent endpoint containing this one.
     * @param relativeUri The URI of this endpoint relative to the
     * <code>parent</code>'s. Prefix <code>./</code> to append a trailing slash
     * to the parent URI if missing.
     */
    public ActionEndpointImpl(Endpoint parent, String relativeUri) {
        super(parent, relativeUri);
    }

    @Override
    public void trigger()
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException {
        executeAndHandle(Request.Post(uri));
    }
}

File: typedrest/src/main/kotlin/net/typedrest/endpoints/rpc/ActionEndpointImpl.kt
package net.typedrest.endpoints.rpc

import net.typedrest.endpoints.*
import net.typedrest.http.uri
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URI

/**
 * RPC endpoint that is invoked with no input or output.
 *
 * @param referrer The endpoint used to navigate to this one.
 * @param relativeUri The URI of this endpoint relative to the referrer's. Add a "./" prefix here to imply a trailing slash on referrer's URI.
 */
open class ActionEndpointImpl(referrer: Endpoint, relativeUri: URI)
    : AbstractRpcEndpoint(referrer, relativeUri), ActionEndpoint {
    /**
     * Creates a new action endpoint.
     *
     * @param referrer The endpoint used to navigate to this one.
     * @param relativeUri The URI of this endpoint relative to the referrer's. Add a "./" prefix here to imply a trailing slash on referrer's URI.
     */
    constructor(referrer: Endpoint, relativeUri: String) :
        this(referrer, URI(relativeUri))

    override fun invoke() =
        execute(Request.Builder().post("".toRequestBody()).uri(uri).build()).close()
}
